File: 0041-first-missing-positive/0041-first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n: int = len(nums)
        containsOne: bool = False
        for i in range(n):
            if nums[i] == 1: containsOne = True
            elif nums[i] <= 0 or nums[i] > n: nums[i] = 1
        
        if not containsOne: return 1

        for i in range(n):
            key: int = abs(nums[i])-1
            if nums[key] > 0: nums[key] *= -1
        
        for i in range(n):
            if nums[i] > 0: return i+1

        return n+1

        


        

File: 0041-first-missing-positive/0041-first-missing-positive.cpp
#include <iostream>
#include <cmath>

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {

        int n = nums.size();
        bool containsOne = false;

        for (int i=0; i<n; i++){
            if (nums[i] == 1) {
                containsOne = true;
            } else if (nums[i] <= 0 || nums[i] > n) {
                nums[i] = 1;
            }
        }

        if (containsOne == false) {
            return 1;
        }

        for (int i=0; i<n; i++) {
            int key = abs(nums[i]) - 1;
            nums[key] = abs(nums[key]) * -1;
        }

        for (int i=0; i<n; i++) {
            if (nums[i] > 0) {
                return i+1;
            }
        }

        return n + 1;
        
    }
};